### PR TITLE
ARROW-15901: [C++] Support flat custom output field names in Substrait

### PR DESCRIPTION
**Pull request by rtpsw**
Replaces https://github.com/apache/arrow/pull/12601

### CODE REVIEW

**Comment by rtpsw** [comment]
@westonpace, I don't know why the errors in Travis CI occurred. Any idea how to resolve them?

File: cpp/src/arrow/compute/exec/options.h [hunk 1]
```diff
@@ -232,10 +232,13 @@ class ARROW_EXPORT SinkNodeConsumer {
...
  public:
-  explicit ConsumingSinkNodeOptions(std::shared_ptr<SinkNodeConsumer> consumer)
-      : consumer(std::move(consumer)) {}
+  explicit ConsumingSinkNodeOptions(std::shared_ptr<SinkNodeConsumer> consumer,
+                                    std::vector<std::string> names = {})
+      : consumer(std::move(consumer)), names(std::move(names)) {}
 
   std::shared_ptr<SinkNodeConsumer> consumer;
+  /// \brief Names to rename the sink's schema fields to
+  std::vector<std::string> names;
```
Comment (vibhatha) [inline cpp/src/arrow/compute/exec/options.h:241]: Nit: May be distinguish the member variables with an underscore?

```c++
std::shared_ptr<SinkNodeConsumer> consumer_;
std::vector<std::string> names_;
```

And change the rest accordingly. Looking into `ConsumingSinkNode`, that is the style adopted.

Comment (westonpace) [reply cpp/src/arrow/compute/exec/options.h:241]: In this case I think the naming scheme is correct.  There are two categories of objects we deal with, passive data objects and more complex stateful objects.  The google style guide likes to use structs for the former and classes for the latter and so it describes this as [structs vs. classes](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes).

In Arrow we are pretty inconsistent about applying `struct` and `class`.  Many of our "options" objects are `class` when they should be `struct` for example.  However, we are pretty consistent about the naming.  For a passive data object all properties should be public and [should not have trailing underscores](https://google.github.io/styleguide/cppguide.html#Variable_Names).

Options objects should always be passive data objects.  Things that travel back and forth across library boundaries should generally be either passive data objects or pure virtual interfaces when possible.  So I do not think a change is needed here.  This options object should be a passive data object and it is following the correct rules for a passive data object.

File: cpp/src/arrow/compute/exec/options.h [hunk 2]
```diff
@@ -232,10 +232,13 @@ class ARROW_EXPORT SinkNodeConsumer {
...
 class ARROW_EXPORT ConsumingSinkNodeOptions : public ExecNodeOptions {
  public:
-  explicit ConsumingSinkNodeOptions(std::shared_ptr<SinkNodeConsumer> consumer)
-      : consumer(std::move(consumer)) {}
+  explicit ConsumingSinkNodeOptions(std::shared_ptr<SinkNodeConsumer> consumer,
+                                    std::vector<std::string> names = {})
+      : consumer(std::move(consumer)), names(std::move(names)) {}
 
   std::shared_ptr<SinkNodeConsumer> consumer;
+  /// \brief Names to rename the sink's schema fields to
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/options.h:240]:
```suggestion
  /// \brief Names to rename the sink's schema fields to
  ///
  /// If specified then names must be provided for all fields
```

File: cpp/src/arrow/engine/substrait/serde.cc [hunk 3]
```diff
@@ -67,15 +67,19 @@ Result<std::vector<compute::Declaration>> DeserializePlan(
...
-      return Status::NotImplemented("substrait::PlanRel with custom output field names");
+      names.assign(plan_rel.root().names().begin(), plan_rel.root().names().end());
     }
-    ARROW_ASSIGN_OR_RAISE(auto decl, FromProto(plan_rel.rel(), ext_set));
 
     // pipe each relation into a consuming_sink node
     auto sink_decl = compute::Declaration::Sequence({
         std::move(decl),
-        {"consuming_sink", compute::ConsumingSinkNodeOptions{consumer_factory()}},
+        {"consuming_sink", compute::ConsumingSinkNodeOptions{consumer_factory(), names}},
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/serde.cc:82]:
```suggestion
        {"consuming_sink", compute::ConsumingSinkNodeOptions{consumer_factory(), std::move(names)}},
```

**Review by westonpace — CHANGES_REQUESTED** [review]
I have a few minor comments but let's also add a test case for this.

This also isn't yet a full implementation of the Substrait spec as it doesn't handle nested fields.  For example, if there are three columns and the first column is a struct column with two fields then the names vector should have 5 items in it.  However, if we want to postpone naming of nested fields to a future PR I think that is fine, just make sure to add a JIRA for it.

File: cpp/src/arrow/compute/exec/options.h (hunk 1 above)
Comment (vibhatha) [reply cpp/src/arrow/compute/exec/options.h]: Ah I see your point. Thank you for the detailed explanation. 👍

**Comment by rtpsw** [comment]
> I have a few minor comments but let's also add a test case for this.

Added.

> This also isn't yet a full implementation of the Substrait spec as it doesn't handle nested fields. For example, if there are three columns and the first column is a struct column with two fields then the names vector should have 5 items in it. However, if we want to postpone naming of nested fields to a future PR I think that is fine, just make sure to add a JIRA for it.

I opted to postpone, and updated the title and the [ARROW-15901](https://issues.apache.org/jira/browse/ARROW-15901) accordingly.

**Comment by rtpsw** [comment]
@westonpace, the AppVeyor failure seems unrelated to the change here. How to proceed?

File: cpp/src/arrow/compute/exec/sink_node.cc [hunk 4]
```diff
@@ -263,18 +263,21 @@ class SinkNode : public ExecNode {
...
  public:
   ConsumingSinkNode(ExecPlan* plan, std::vector<ExecNode*> inputs,
-                    std::shared_ptr<SinkNodeConsumer> consumer)
+                    std::shared_ptr<SinkNodeConsumer> consumer,
+                    std::vector<std::string> names)
       : ExecNode(plan, std::move(inputs), {"to_consume"}, {},
                  /*num_outputs=*/0),
-        consumer_(std::move(consumer)) {}
+        consumer_(std::move(consumer)),
+        names_(names) {}
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/sink_node.cc:271]:
```suggestion
        names_(std::move(names)) {}
```

**Review by westonpace — APPROVED** [review]
The appveyor failure seems unrelated.  One small suggestion and I think this is good to go.

**Comment by rtpsw** [comment]
@westonpace, this has been pending for a while. How to proceed?

**Comment by westonpace** [comment]
I think this is ready to go.  If @vibhatha has any more thoughts he can of course still mention them and we can address in a follow-up.

**Pull request merged**